In [136]:
from gurobipy import GRB
import gurobipy as gb
import pandas as pd
import numpy as np

In [137]:
# Create the optimization model
model = gb.Model("Question 2: S&P500")

In [138]:
investment_df = pd.read_csv(r"C:\Users\gabri\Downloads\sp500_data.csv")

In [139]:
investment_df.head()

Ticker symbol                   Company                  GICS Sector   
0           AEE               Ameren Corp                    Utilities  \
1           AXP       American Express Co                   Financials   
2             T                  AT&T Inc  Telecommunications Services   
3           AVP             Avon Products             Consumer Staples   
4           BFB  Brown-Forman Corporation             Consumer Staples   

  Location of Headquarters  Price  PercentReturn  
0      St. Louis, Missouri  31.26           4.83  
1       New York, New York  49.23           1.53  
2            Dallas, Texas  30.09           5.82  
3       New York, New York  18.31           5.26  
4     Louisville, Kentucky  82.28           1.74

In [140]:
investment_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 67 entries, 0 to 66
Data columns (total 6 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Ticker symbol             67 non-null     object 
 1   Company                   67 non-null     object 
 2   GICS Sector               67 non-null     object 
 3   Location of Headquarters  67 non-null     object 
 4   Price                     67 non-null     float64
 5   PercentReturn             67 non-null     float64
dtypes: float64(2), object(4)
memory usage: 3.3+ KB


In [141]:
investment_df.columns

Index(['Ticker symbol', 'Company', 'GICS Sector', 'Location of Headquarters',
       'Price', 'PercentReturn'],
      dtype='object')

In [142]:
rate_of_return = investment_df['PercentReturn'].values

In [143]:
rate_of_return

array([ 4.83,  1.53,  5.82,  5.26,  1.74,  1.89,  4.22,  3.49,  1.47,
        2.69,  2.02,  3.62,  2.14,  0.9 ,  1.67,  3.71,  7.21,  2.17,
        4.32,  4.55,  0.15,  2.22,  1.86, 14.56,  2.39,  3.79,  1.29,
        4.63,  3.55,  1.86,  2.44,  3.48,  3.01,  1.56,  4.68,  3.67,
        0.16,  2.79,  1.03,  4.46,  3.8 ,  3.08,  2.33,  2.58,  0.94,
        4.9 ,  5.32,  3.1 ,  1.4 ,  4.42,  4.36,  4.59,  3.15,  0.27,
        1.57,  2.88,  1.48,  4.49,  2.23,  2.77,  1.75,  2.84,  0.76,
        2.63,  4.99,  2.44,  4.21])

In [144]:
x = model.addVars(67, lb=0, vtype=GRB.CONTINUOUS, name="Money Invested")

In [145]:
# Define the objective function
objective = gb.quicksum((rate_of_return[i] / 100) * x[i] for i in range(67))

# Set the objective function in the model
model.setObjective(objective, GRB.MAXIMIZE)

In [146]:
money_invested = 10000000

In [147]:
total_invested = model.addConstr(gb.quicksum(x[j] for j in range(67)) == money_invested, name="Investment constraint")

In [148]:
for j in range(67):
    model.addConstr(x[j] <= 600000, f"Diversification Constraint_{j}")

In [149]:
investment_df['GICS Sector'].unique()

array(['Utilities', 'Financials', 'Telecommunications Services',
       'Consumer Staples', 'Industrials', 'Consumer Discretionary',
       'Energy', 'Information Technology', 'Health Care', 'Materials'],
      dtype=object)

In [150]:
sectors = investment_df['GICS Sector'].values
sectors

array(['Utilities', 'Financials', 'Telecommunications Services',
       'Consumer Staples', 'Consumer Staples', 'Industrials',
       'Consumer Discretionary', 'Consumer Staples',
       'Consumer Discretionary', 'Consumer Staples', 'Consumer Staples',
       'Energy', 'Industrials', 'Energy', 'Financials', 'Utilities',
       'Industrials', 'Industrials', 'Utilities', 'Utilities', 'Energy',
       'Energy', 'Consumer Discretionary', 'Telecommunications Services',
       'Consumer Discretionary', 'Industrials', 'Consumer Discretionary',
       'Financials', 'Consumer Staples', 'Information Technology',
       'Financials', 'Health Care', 'Financials', 'Financials',
       'Financials', 'Financials', 'Information Technology',
       'Consumer Discretionary', 'Health Care', 'Health Care',
       'Information Technology', 'Information Technology', 'Materials',
       'Utilities', 'Information Technology', 'Financials', 'Utilities',
       'Consumer Staples', 'Health Care', 'Utilities', 'U

In [151]:
telecommunications_indices = []
for tele in sectors:
    if tele in ['Telecommunications Services']:
        telecommunications_indices.append(1)
    else:
        telecommunications_indices.append(0)

In [152]:
telecommunications_indices

[0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0]

In [153]:
model.addConstr((gb.quicksum(telecommunications_indices[i]*x[i] for i in range(67))) <= 500000, "Telecommunications Constraint")

<gurobi.Constr *Awaiting Model Update*>

In [154]:
IT_indices = []
for job in sectors:
    if job in ['Information Technology']:
        IT_indices.append(1)
    else:
        IT_indices.append(0)

In [155]:
IT_indices

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0]

In [156]:
model.addConstr((gb.quicksum(IT_indices[i]*x[i] for i in range(67))) >= 0.75*(gb.quicksum(telecommunications_indices[i]*x[i] for i in range(67))), "IT Constraint")

<gurobi.Constr *Awaiting Model Update*>

In [157]:
investment_df['GICS Sector'].unique()

array(['Utilities', 'Financials', 'Telecommunications Services',
       'Consumer Staples', 'Industrials', 'Consumer Discretionary',
       'Energy', 'Information Technology', 'Health Care', 'Materials'],
      dtype=object)

In [158]:
CD_indices = []
for job in sectors:
    if job in ['Consumer Discretionary']:
        CD_indices.append(1)
    else:
        CD_indices.append(0)

CD_indices

[0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 1]

In [159]:
CS_indices = []
for job in sectors:
    if job in ['Consumer Staples']:
        CS_indices.append(1)
    else:
        CS_indices.append(0)

CS_indices

[0,
 0,
 0,
 1,
 1,
 0,
 0,
 1,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0]

In [160]:
model.addConstr(gb.quicksum(CD_indices[j] * x[j] for j in range(67)) - gb.quicksum(CS_indices[j] * x[j] for j in range(67)) <= 200000, name="Absolute Value constraint")
model.addConstr(gb.quicksum(CS_indices[j] * x[j] for j in range(67)) - gb.quicksum(CD_indices[j] * x[j] for j in range(67)) <= 200000, name="Absolute Value constraint negative")

<gurobi.Constr *Awaiting Model Update*>

In [161]:
energy_indices = []
for job in sectors:
    if job in ['Energy']:
        energy_indices.append(1)
    else:
        energy_indices.append(0)

energy_indices

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0]

In [162]:
energy = model.addConstr((gb.quicksum(energy_indices[i]*x[i] for i in range(67))) >= 1000000, "Energy Constraint")

In [163]:
investment_df

Ticker symbol                    Company                  GICS Sector   
0            AEE                Ameren Corp                    Utilities  \
1            AXP        American Express Co                   Financials   
2              T                   AT&T Inc  Telecommunications Services   
3            AVP              Avon Products             Consumer Staples   
4            BFB   Brown-Forman Corporation             Consumer Staples   
..           ...                        ...                          ...   
62             X  United States Steel Corp.                    Materials   
63           UTX        United Technologies                  Industrials   
64            VZ     Verizon Communications  Telecommunications Services   
65           WMT            Wal-Mart Stores             Consumer Staples   
66           WHR            Whirlpool Corp.       Consumer Discretionary   

             Location of Headquarters  Price  PercentReturn  
0                 St. Louis, Missouri  31.26           4.83  
1                  New York, New York  49.23           1.53  
2                       Dallas, Texas  30.09           5.82  
3                  New York, New York  18.31           5.26  
4                Louisville, Kentucky  82.28           1.74  
..                                ...    ...            ...  
62           Pittsburgh, Pennsylvania  28.96           0.76  
63              Hartford, Connecticut  77.78           2.63  
64                 New York, New York  37.79           4.99  
65              Bentonville, Arkansas  61.39           2.44  
66  Benton Charter Township, Michigan  54.36           4.21  

[67 rows x 6 columns]

In [164]:
investment_df['Location of Headquarters'].unique()

array(['St. Louis, Missouri', 'New York, New York', 'Dallas, Texas',
       'Louisville, Kentucky', 'Eden Prairie, Minnesota',
       'Bethpage, New York', 'Camden, New Jersey', 'Atlanta, Georgia',
       'Houston, Texas', 'Riverwoods, Illinois', 'Richmond, Virginia',
       'Chicago, Illinois', 'Downers Grove, Illinois',
       'Detroit, Michigan', 'Charlotte, North Carolina', 'Irving, Texas',
       'Dearborn, Michigan', 'Stamford, Connecticut', 'McLean, Virginia',
       'Fairfield, Connecticut', 'Milwaukee, Wisconsin',
       'Long Beach, California', 'Pittsburgh, Pennsylvania',
       'Palo Alto, California', 'New Brunswick, New Jersey',
       'Cleveland, Ohio', 'New Hyde Park, New York', 'Buffalo, New York',
       'Harrison, New York', 'Oak Brook, Illinois',
       'San Francisco, California', 'Whitehouse Station, New Jersey',
       'Chandler, Arizona', 'Redmond, Washington', 'Denver, Colorado',
       'Tulsa, Oklahoma', 'Redwood Shores, California',
       'Bridgeport, Connec

In [165]:
location = investment_df['Location of Headquarters'].values
location

array(['St. Louis, Missouri', 'New York, New York', 'Dallas, Texas',
       'New York, New York', 'Louisville, Kentucky',
       'Eden Prairie, Minnesota', 'Bethpage, New York',
       'Camden, New Jersey', 'New York, New York', 'Atlanta, Georgia',
       'Atlanta, Georgia', 'Houston, Texas', 'Houston, Texas',
       'Houston, Texas', 'Riverwoods, Illinois', 'Richmond, Virginia',
       'Chicago, Illinois', 'Downers Grove, Illinois',
       'Detroit, Michigan', 'Charlotte, North Carolina', 'Houston, Texas',
       'Irving, Texas', 'Dearborn, Michigan', 'Stamford, Connecticut',
       'McLean, Virginia', 'Fairfield, Connecticut',
       'Milwaukee, Wisconsin', 'Long Beach, California',
       'Pittsburgh, Pennsylvania', 'Palo Alto, California',
       'Atlanta, Georgia', 'New Brunswick, New Jersey',
       'New York, New York', 'Cleveland, Ohio', 'New Hyde Park, New York',
       'Buffalo, New York', 'Harrison, New York', 'Oak Brook, Illinois',
       'San Francisco, California', 'White

In [166]:
new_york_indices = []
for place in location:
    if place in ['New York, New York']:
        new_york_indices.append(1)
    else:
        new_york_indices.append(0)

new_york_indices

[0,
 1,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0]

In [167]:
NY = model.addConstr((gb.quicksum(new_york_indices[i]*x[i] for i in range(67))) >= 300000, "Location Constraint")

In [168]:
# Optimally solve the problem
model.optimize()

Gurobi Optimizer version 11.0.0 build v11.0.0rc2 (win64 - Windows 11+.0 (22631.2))

CPU model: Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 6 physical cores, 12 logical processors, using up to 12 threads

Optimize a model with 74 rows, 67 columns and 196 nonzeros
Model fingerprint: 0x4b090aa3
Coefficient statistics:
  Matrix range     [8e-01, 1e+00]
  Objective range  [2e-03, 1e-01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [2e+05, 1e+07]
Presolve removed 68 rows and 16 columns
Presolve time: 0.00s
Presolved: 6 rows, 52 columns, 92 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    1.4560000e+06   3.474937e+06   0.000000e+00      0s
       6    5.1346000e+05   0.000000e+00   0.000000e+00      0s

Solved in 6 iterations and 0.01 seconds (0.00 work units)
Optimal objective  5.134600000e+05


In [169]:
# Print the decision variables
print(model.printAttr('X'))


    Variable            X 
-------------------------
Money Invested[0]       600000 
Money Invested[3]       600000 
Money Invested[6]       400000 
Money Invested[11]       600000 
Money Invested[16]       600000 
Money Invested[19]       600000 
Money Invested[21]       400000 
Money Invested[23]       500000 
Money Invested[27]       600000 
Money Invested[34]       600000 
Money Invested[39]       600000 
Money Invested[40]       375000 
Money Invested[45]       600000 
Money Invested[46]       600000 
Money Invested[49]       600000 
Money Invested[50]       525000 
Money Invested[51]       600000 
Money Invested[57]       600000 
None


In [170]:
# Value of the objective function
print("Expected 1-year Return: ", round(model.objVal, 2))

Expected 1-year Return:  513460.0


In [171]:
total = sum(new_york_indices)
total

6

a) 6

b) 67

c) A decision variable represents the amount of money to be invested in a company. It is i = 1 through 67. Constraint: for i = 1, x_i <= 600000

In [172]:
invested_money = model.getAttr('X', x)
invested_money

{0: 600000.0,
 1: 0.0,
 2: 0.0,
 3: 600000.0,
 4: 0.0,
 5: 0.0,
 6: 400000.0,
 7: 0.0,
 8: 0.0,
 9: 0.0,
 10: 0.0,
 11: 600000.0,
 12: 0.0,
 13: 0.0,
 14: 0.0,
 15: 0.0,
 16: 600000.0,
 17: 0.0,
 18: 0.0,
 19: 600000.0,
 20: 0.0,
 21: 400000.0,
 22: 0.0,
 23: 500000.0,
 24: 0.0,
 25: 0.0,
 26: 0.0,
 27: 600000.0,
 28: 0.0,
 29: 0.0,
 30: 0.0,
 31: 0.0,
 32: 0.0,
 33: 0.0,
 34: 600000.0,
 35: 0.0,
 36: 0.0,
 37: 0.0,
 38: 0.0,
 39: 600000.0,
 40: 375000.0,
 41: 0.0,
 42: 0.0,
 43: 0.0,
 44: 0.0,
 45: 600000.0,
 46: 600000.0,
 47: 0.0,
 48: 0.0,
 49: 600000.0,
 50: 525000.0,
 51: 600000.0,
 52: 0.0,
 53: 0.0,
 54: 0.0,
 55: 0.0,
 56: 0.0,
 57: 600000.0,
 58: 0.0,
 59: 0.0,
 60: 0.0,
 61: 0.0,
 62: 0.0,
 63: 0.0,
 64: 0.0,
 65: 0.0,
 66: 0.0}

In [173]:
new_york_investment = [new_york_indices[i] * invested_money[i] for i in range(67)]
new_york_investment

[0.0,
 0.0,
 0.0,
 600000.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0]

In [174]:
total_NY_investment = sum(new_york_investment)
total_NY_investment

600000.0

d) $600,000 invested in NYC companies.

e) OFV = $513460

In [175]:
print("")
print(f"Sensitivity Information for Land Capacity Constraint {energy.pi:.2f}:")
print("(LHS, RHS, Slack): ", (model.getRow(energy).getValue(), energy.RHS, energy.slack))
print("Shadow Price: ", energy.pi)
print("Range of Feasibility: ", (energy.SARHSUp, energy.SARHSLow))


Sensitivity Information for Land Capacity Constraint -0.02:
(LHS, RHS, Slack):  (1000000.0, 1000000.0, 0.0)
Shadow Price:  -0.0214
Range of Feasibility:  (1200000.0, 925000.0)


In [176]:
print("")
print(f"Sensitivity Information for Land Capacity Constraint {NY.pi:.2f}:")
print("(LHS, RHS, Slack): ", (model.getRow(NY).getValue(), NY.RHS, NY.slack))
print("Shadow Price: ", NY.pi)
print("Range of Feasibility: ", (NY.SARHSUp, NY.SARHSLow))


Sensitivity Information for Land Capacity Constraint 0.00:
(LHS, RHS, Slack):  (600000.0, 300000.0, -300000.0)
Shadow Price:  0.0
Range of Feasibility:  (600000.0, -inf)


In [177]:
print("")
print(f"Sensitivity Information for Land Capacity Constraint {total_invested.pi:.2f}:")
print("(LHS, RHS, Slack): ", (model.getRow(total_invested).getValue(), total_invested.RHS, total_invested.slack))
print("Shadow Price: ", total_invested.pi)
print("Range of Feasibility: ", (total_invested.SARHSUp, total_invested.SARHSLow))


Sensitivity Information for Land Capacity Constraint 0.04:
(LHS, RHS, Slack):  (10000000.0, 10000000.0, 0.0)
Shadow Price:  0.0436
Range of Feasibility:  (10075000.0, 9475000.0)


f) I guess relative to what is being solved the shadow price is a percentage. Since it is negative, provided the lowering is in the range of feasibility to ensure the shadow price can properly account for it, lowering the number invested increases the OFV, so do it.

In [178]:
investment_df.head(10)

Ticker symbol                    Company                  GICS Sector   
0           AEE                Ameren Corp                    Utilities  \
1           AXP        American Express Co                   Financials   
2             T                   AT&T Inc  Telecommunications Services   
3           AVP              Avon Products             Consumer Staples   
4           BFB   Brown-Forman Corporation             Consumer Staples   
5          CHRW   C. H. Robinson Worldwide                  Industrials   
6           CVC  Cablevision Systems Corp.       Consumer Discretionary   
7           CPB              Campbell Soup             Consumer Staples   
8           CBS                  CBS Corp.       Consumer Discretionary   
9            KO              Coca Cola Co.             Consumer Staples   

  Location of Headquarters  Price  PercentReturn  
0      St. Louis, Missouri  31.26           4.83  
1       New York, New York  49.23           1.53  
2            Dallas, Texas  30.09           5.82  
3       New York, New York  18.31           5.26  
4     Louisville, Kentucky  82.28           1.74  
5  Eden Prairie, Minnesota  67.63           1.89  
6       Bethpage, New York  14.16           4.22  
7       Camden, New Jersey  31.61           3.49  
8       New York, New York  28.64           1.47  
9         Atlanta, Georgia  67.90           2.69

In [179]:
investment_df.iloc[10:21]

Ticker symbol                      Company       GICS Sector   
10           CCE        Coca-Cola Enterprises  Consumer Staples  \
11           COP               ConocoPhillips            Energy   
12           CBE            Cooper Industries       Industrials   
13            DO    Diamond Offshore Drilling            Energy   
14           DFS  Discover Financial Services        Financials   
15             D           Dominion Resources         Utilities   
16           RRD      Donnelley (R.R.) & Sons       Industrials   
17           DOV                  Dover Corp.       Industrials   
18           DTE               DTE Energy Co.         Utilities   
19           DUK                  Duke Energy         Utilities   
20            EP                El Paso Corp.            Energy   

     Location of Headquarters  Price  PercentReturn  
10           Atlanta, Georgia  26.59           2.02  
11             Houston, Texas  70.61           3.62  
12             Houston, Texas  60.16           2.14  
13             Houston, Texas  61.31           0.90  
14       Riverwoods, Illinois  27.83           1.67  
15         Richmond, Virginia  50.31           3.71  
16          Chicago, Illinois  12.20           7.21  
17    Downers Grove, Illinois  60.47           2.17  
18          Detroit, Michigan  52.96           4.32  
19  Charlotte, North Carolina  21.14           4.55  
20             Houston, Texas  26.75           0.15

g) No because it is 3-2.02, which is 0.98, which is within the allowable increase. This means that the optimal portfolio does not change.

h) Since the shadow price is positive, an increase in the RHS value outside of the allowable increase results in the optimal funds changing, so do it.

i) Since it is outside the allowable increase, using the equation 4.32% × 50000 − 1.5% × 50000 will result in a positive number which is the increase in profit, do it.

j) 6 - (1*0 + 1*4.36 - 1*2.14) will be positive, so do it.